pyro/ops/stats.py

```diff
@@ -0,0 +1,208 @@
+    ac = ac / ac[..., :1]
+    return ac.transpose(dim, -1)
+
+
+def autocovariance(input, dim=0):
+    """
+    Computes the autocovariance of samples at dimension `dim`.
+
+    :param torch.Tensor input: the input tensor.
+    :param int dim: the dimension to calculate autocorrelation.
+    :returns torch.Tensor: autocorrelation of `input`.
+    """
+    return autocorrelation(input) * input.var(dim, unbiased=False, keepdim=True)
+
+
+def _cummin(input):
+    """
+    Computes cummulative minimum of input at dimension `dim=0`.
+    """
+    # FIXME: is there a better trick to find accumulate min of a sequence?
+    N = input.size(0)
+    input_tril = input.unsqueeze(0).repeat((N,) + (1,) * input.dim())
+    triu_mask = input.new_ones(N, N).triu(diagonal=1).reshape((N, N) + (1,) * (input.dim() - 1))
+    triu_mask = triu_mask.expand((N, N) + input.shape[1:]) > 0.5
+    input_tril.masked_fill_(triu_mask, input.max())
+    return input_tril.min(dim=1)[0]
+
+
+def effective_sample_size(input, chain_dim=0, sample_dim=1):
+    """
+    Computes effective sample size of input.
+
+    Reference:
+    [1] `Introduction to Markov Chain Monte Carlo`,
+        Charles J. Geyer
+    [2] `Stan Reference Manual version 2.18`,
+        Stan Development Team
+
+    :param torch.Tensor input: the input tensor.
+    :param int chain_dim: the chain dimension.
+    :param int sample_dim: the sample dimension.
+    :returns torch.Tensor: effective sample size of `input`.
+    """
+    assert input.dim() >= 2
+    assert input.size(sample_dim) >= 2
+    assert input.size(chain_dim) >= 2
+    # change input.shape to 1 x 1 x input.shape
+    # then transpose sample_dim with 0, chain_dim with 1
+    sample_dim = input.dim() + sample_dim if sample_dim < 0 else sample_dim
+    chain_dim = input.dim() + chain_dim if chain_dim < 0 else chain_dim
+    assert chain_dim != sample_dim
+    input = input.reshape((1, 1) + input.shape)
+    input = input.transpose(0, sample_dim + 2).transpose(1, chain_dim + 2)
+
+    N, C = input.size(0), input.size(1)
+    # find autocovariance for each chain at lag k
+    gamma_k_c = autocovariance(input, dim=0)  # N x C x sample_shape
+
+    # find autocorrelation at lag k (from Stan reference)
+    var_within, var_estimator = _compute_chain_variance_stats(input)
+    rho_k = (var_estimator - var_within + gamma_k_c.mean(dim=1)) / var_estimator
+    rho_k[0] = 1  # correlation at lag 0 is always 1
+
+    # initial positive sequence (formula 1.18 in [1]) applied for autocorrelation
+    Rho_k = rho_k if N % 2 == 0 else rho_k[:-1]
+    Rho_k = Rho_k.reshape((N // 2, 2) + Rho_k.shape[1:]).sum(dim=1)
+
+    # separate the first index
+    Rho_init = Rho_k[0]
+
+    if Rho_k.size(0) > 1:
+        # Theoretically, Rho_k is positive, but due to noise of correlation computation,
+        # Rho_k might not be positive at some point. So we need to truncate (ignore first index).
+        Rho_positive = Rho_k[1:].clamp(min=0)
```

Comment:
Maybe I'm misunderstanding, but I would imagine that this will turn the entire sequence after the occurrence of a negative element to 0, which won't be the case if you were to remove the elements.

Reply:
Makes sense, thanks for explaining @fehiepsi.